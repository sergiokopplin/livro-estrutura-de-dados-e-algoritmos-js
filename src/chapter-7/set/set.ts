export class Set {
  items: any;

  constructor() {
    this.items = {};
  }

  has(element: any): boolean {
    return Object.prototype.hasOwnProperty.call(this.items, element);
  }

  add(element: any): boolean {
    if (!this.has(element)) {
      this.items[element] = element;

      return true;
    }

    return false;
  }

  delete(element: any): boolean {
    if (this.has(element)) {
      delete this.items[element];
      return true;
    }

    return false;
  }

  clear(): void {
    this.items = {};
  }

  size(): number {
    return Object.keys(this.items).length;
  }

  values(): any[] {
    return Object.values(this.items);
  }
}
