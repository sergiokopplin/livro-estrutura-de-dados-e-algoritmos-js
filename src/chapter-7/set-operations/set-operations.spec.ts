import { Set } from '../set/set';
import { SetOperations } from './set-operations';

describe('SetOperations', () => {
  test('Should construct correctly', () => {
    const setOperations = new SetOperations();

    expect(setOperations).toMatchSnapshot();
  });

  test('Should Union', () => {
    const setA = new SetOperations();

    setA.add(1);
    setA.add(2);
    setA.add(3);

    const setB = new Set();

    setB.add(3);
    setB.add(4);
    setB.add(5);
    setB.add(6);

    expect(setA.union(setB)).toEqual([1, 2, 3, 4, 5, 6]);
  });

  test('Should Intersect', () => {
    const setA = new SetOperations();

    setA.add(1);
    setA.add(2);
    setA.add(3);

    const setB = new Set();

    setB.add(2);
    setB.add(3);
    setB.add(4);

    expect(setA.intersection(setB)).toEqual([2, 3]);
  });

  test('Should Difference', () => {
    const setA = new SetOperations();

    setA.add(1);
    setA.add(2);
    setA.add(3);

    const setB = new Set();

    setB.add(2);
    setB.add(3);
    setB.add(4);

    expect(setA.difference(setB)).toEqual([1]);
  });

  test('Should calculate subsetOf', () => {
    const setA = new SetOperations();

    setA.add(1);
    setA.add(2);

    const setB = new SetOperations();

    setB.add(1);
    setB.add(2);
    setB.add(3);

    const setC = new SetOperations();

    setC.add(2);
    setC.add(3);
    setC.add(4);

    expect(setA.isSubsetOf(setB)).toBe(true);
    expect(setB.isSubsetOf(setC)).toBe(false);
  });
});
