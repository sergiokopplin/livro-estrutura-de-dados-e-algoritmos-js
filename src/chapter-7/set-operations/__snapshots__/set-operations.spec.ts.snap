// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetOperations Should construct correctly 1`] = `
SetOperations {
  "items": {},
}
`;
